License header check

YanxuanLiu opened the pull request:
Add license header check common action.

Users need to config the excluded files pattern in their source code repo.

Update 11.14:

1. Refactored the common action into a simpler structure. Removed Dockerfile and python script, run bash in workflow directly.
2. Replaced excluded file pattern with included file patterns (wildcard character). I listed all file types of our repos, declaring included file patterns could be a easier way, since we only care about several file types.
3. Now the action will check the year in copyright, it would fail if copyright was expired.

pxLi commented:
thanks! I will review this later

gerashegalov commented:
I am fine if this is something that already works. However, I am somewhat surprised to see the size of this PR for something that boils down to checking if the output of `git grep  -L 'Copyright (c) .* NVIDIA'` matches exclusion patterns.

Do we really need a dedicated Dockerfile ?

pxLi commented on license-header-check/license-header-check:
```diff
@@ -0,0 +1,105 @@
+    return files
+
+def filterFiles(files, excludes):
+    excludes = excludes.split('\n')
```
how should people pass the patterns? a file or a string? use '\n' to split is not safe IMHO, the format checking should be strict. otherwise, it could match unrelated ones.

pxLi reviewed (commented):
please also provide more example runs (different patterns, success and fail runs) and README for how to use this action

pxLi commented:
> Do we really need a dedicated Dockerfile ?

its not a must, we can install deps within the action steps itself. but the dockerfile could help people port the code to somewhere else easlity if needed

pxLi commented:
> However, I am somewhat surprised to see the size of this PR for something that boils down to checking if the output of git grep  -L 'Copyright (c) .* NVIDIA' matches exclusion patterns.

Yes, I agree. @YanxuanLiu this change seems overly complex for a simple check. IMHO, a basic bash step would be enough to cover the case.

YanxuanLiu commented:
> > However, I am somewhat surprised to see the size of this PR for something that boils down to checking if the output of git grep  -L 'Copyright (c) .* NVIDIA' matches exclusion patterns.
> 
> Yes, I agree. @YanxuanLiu this change seems overly complex for a simple check. IMHO, a basic bash step would be enough to cover the case.

OK, let me change to a simpler way, since our repos are not huge ones.

YanxuanLiu commented:
> I am fine if this is something that already works. However, I am somewhat surprised to see the size of this PR for something that boils down to checking if the output of `git grep -L 'Copyright (c) .* NVIDIA'` matches exclusion patterns.
> 
> Do we really need a dedicated Dockerfile ?

Thanks Gera, I've converted this PR to Draft, will refactor it in a simpler way.

YanxuanLiu commented:
Hi @pxLi @gerashegalov , I refactored the common action, will attach more test runs later.

pxLi commented on license-header-check/action.yml:
```diff
@@ -0,0 +1,68 @@
+        IFS="," read -r -a FILE_PATTERN_ARRAY <<< ${{ inputs.included_file_patterns }}
+        PATTERNS=""
+        for PATTERN in "${FILE_PATTERN_ARRAY[@]}"; do
+            PATTERNS+="$(echo "$PATTERN" | xargs -I{} echo "{}") "
```
what does this for?

can we just put the patterns array in below FILES loop? and maybe simplify the find+grep combination if possible like
```
        for FILE in $FILES; do
            echo "Detected changed file: $FILE"
            for PATTERN in "${FILE_PATTERN_ARRAY[@]}"; do
                if [[ "$FILE" =~ $PATTERN ]]; then
```

YanxuanLiu replied:
This is to trim the patterns in pattern array, I'll simplify the step here, maybe merge with `read` command. 

`[[ "$FILE" =~ $PATTERN ]]`  cannot work as expected with `wildcard`, I think `wildcard` as pattern could be clearer for uses compared with `regex`.

YanxuanLiu replied:
Replaced `find` with `[[ $(basename $FILE) == $PATTERN ]]` as discussed in slack

pxLi commented on license-header-check/action.yml:
```diff
@@ -0,0 +1,68 @@
+            for PATTERN in $PATTERNS; do
+                if find "$(dirname $FILE)" -type f -name "$PATTERN" | grep -wq "$FILE"; then
+                    echo "Checking $FILE"
+                    if !(grep -E "$LICENSE_PATTERN" "$FILE"); then
```
nit: `grep -Eq` for quiet check only

YanxuanLiu replied:
Replaced `find` with `[[ $(basename $FILE) == $PATTERN ]]` as discussed in slack

pxLi commented on license-header-check/action.yml:
```diff
@@ -0,0 +1,68 @@
+        echo "--------- RESULT ---------"
+        if [ ! -z "$NO_LICENSE_FILES" ]; then
+            echo "Following files missed copyright/license header or expired: "
+            for FILE in $NO_LICENSE_FILES; do
```
nit: no need loop here

YanxuanLiu replied:
Replaced with a one-line echo

pxLi reviewed (commented):
This looks much better

gerashegalov commented on license-header-check/action.yml:
```diff
@@ -0,0 +1,68 @@
+        done
+
+        # Get changed files
+        FILES=$(git diff --name-only ${{ github.event.pull_request.base.sha }} ${{ github.event.pull_request.head.sha }})
```
I wonder if we should have an optional check for all files in the repo (not just diff) showing if there are missed files from before this check has been introduced?

pxLi replied:
this could be a one-time run before we enable the check for each repo, and then we just need to detect modified files only

YanxuanLiu commented on license-header-check/action.yml:
```diff
@@ -0,0 +1,68 @@
+      shell: bash
+      run: |
+        CURRENT_YEAR=$(date +%Y)
+        LICENSE_PATTERN="Copyright \(c\) .*?${CURRENT_YEAR}, NVIDIA CORPORATION."
```
Yes, in current rules, it will fail if the file is renamed only. We optimize in 2 options:

1. `git diff --name-only --diff-filter=R` ignore all renamed files.
2. `git diff --name-only --diff-filter=AM` detect new added and modified files. If the content of renamed file is modified, we still check the header.

YanxuanLiu replied:
Updated the action with option 3. Renamed files with content changes will be checked.

YanxuanLiu commented:
[Test run:](https://github.com/YanxuanLiu/spark-rapids-ml/actions/runs/11853838261/job/33034750770):

- Deleted files: ignore
- Renamed files (without content change): ignore
- Renamed files (with content change): check
- Added/Modified files: check

YanxuanLiu commented:
@gerashegalov Added arg for excluded files patterns. Could you help to review again. Then I will start from JNI repo to enable the check. Thanks!

gerashegalov reviewed (approved):
LGTM

pxLi reviewed (approved):
thanks! lets try enable it in JNI first to verify the permission setup and usages first 👍
